### ISSUE TITLE
Numerically stable computation of invariant mass

### ISSUE THREAD

**Issue by HDembinski** [scikit-hep/vector#188]
I recently developed a formula to compute the invariant mass for a two-body decay in a numerically stable way from the momenta of the daughters and their assumed masses. The algorithm is described and benchmarked here:
https://github.com/HDembinski/essays/blob/master/Numerically%20stable%20calculation%20of%20invariant%20mass.ipynb
The numerical instability arises from the subtractive cancellation in this formula

![image](https://user-images.githubusercontent.com/2631586/165310291-06f8ecca-9d66-407d-b512-8ba7e2426118.png)

for m1,m2 << |p1|, |p2|.

I am not sure whether this fits in here, but I wanted to advertise it. As you can see in my notebook, my formula improves the accuracy dramatically over the naive formula. It further allows one to perform the whole calculation in single precision instead of double precision, which frees CPU registers and should accelerate code that benefits from SIMD instructions. Single precision is sufficiently accurate to describe the resolution of experimental data, we only habitually convert to double precision because we usually do not use numerically stable formulas.

**Comment by jpivarski** [comment]
That would be great to add! Worse than not being numerically stable, Vector doesn't yet have a function for computing the invariant mass of two 4-vectors independently of the other three components. That is, users are currently computing

```python
(vector1 + vector2).mass
```

which computes all four components of `vector1 + vector2`, uses them in the computation of `mass`, and then throws away the temporaries. We ought to have a function that just computes invariant mass and nothing else, but we don't (yet). I had been thinking of adding this formula:

```python
mass = sqrt(2*pt1*pt2*(cosh(eta1 - eta2) - cos(phi1 - phi2)))
```

but that has the numerically unstable subtraction (as well as an approximation that the mass components of `vector1` and `vector2` are zero). It is the case that `sqrt` sometimes encounters negative values due to round-off error, and I see that your implementation is guaranteed to not produce negative values.

----------------

Adding a function to Vector is a bit of a process. It supports many backends by duck-typing: the source code in the `compute` submodule is generic and takes `lib` as an argument, where `lib` may be `numpy` or some other library with the same functions. [This test](https://github.com/scikit-hep/vector/blob/main/tests/test_compute_features.py) acts as a gatekeeper, warning us if any compute functions use features we haven't yet determined are "safe." (It's a whitelist that we can add to.)

These functions are currently being used on Python scalars, NumPy arrays, Awkward Arrays, and in Numba (all of which use `lib = numpy`, though Python objects could have used `lib = math`), but the idea is that we should be able to extend to CuPy (`lib = cupy`), autodifferentiable JAX (`lib = jax.numpy`), etc. I think a tour de force would be to show that it can even be used on symbolic math (`lib = sympy` or a SymPy wrapper) because such weak assumptions are being made.

Another dimension that scales Vector functions is that we have a set of 2 × 3 × 2 = 12 coordinate systems:

   * azimuthal coordinates may be "x, y" or "rho, phi": 2 systems
   * longitudinal coordinate may be "z" or "theta" or "eta": 3 systems
   * temporal coordinate may be "t" (Cartesian) or "tau" (proper time): 2 systems

That is, the azimuthal coordinates are independent of all others, longitudinal components depend only on the longitudinal direction and azimuthal, and temporal components depend on all three. This makes it impossible to use "r" (3D spatial distance) as a stored coordinate, but everything else that we typically want is available (e.g. proper time with any combination of Cartesian, polar, and pseudorapidity).

Unfortunately, that blows up the number of formulas that in principle need to be written in the compute submodule, especially for 4D vectors, especially for multiple inputs (an invariant mass calculation has two vector inputs). For 2D and to a lesser extent, 3D, we often write out specialized formulas when they're available in a combination of coordinate systems, but most 4D calculations either use 2D and 3D functions directly or do coordinate transformations before computing if necessary.

Here's an example: the `dot` function in 2D

https://github.com/scikit-hep/vector/blob/b9c60a8320b39812354c48182281480087565f5e/src/vector/_compute/planar/dot.py#L28-L41

has an explicit formula when both vectors are Cartesian or when both vectors are polar, but if one is Cartesian and the other is polar, it converts to Cartesian. Here, `x` and `y` are modules like the `dot` module,

https://github.com/scikit-hep/vector/blob/b9c60a8320b39812354c48182281480087565f5e/src/vector/_compute/planar/dot.py#L16

and the `x.rhophi` and `y.rhophi` functions convert "rho, phi" coordinates into "x" and "y" respectively. The naming convention is absolutely consistent.

Methods on vectors look up compute functions through a dict, which is keyed on combinations of coordinate systems and includes signature information (this function returns `float` on individual vectors):

https://github.com/scikit-hep/vector/blob/b9c60a8320b39812354c48182281480087565f5e/src/vector/_compute/planar/dot.py#L44-L49

Some of the 3D and 4D functions use the lookup programmatically to generate all of the coordinate system combinations. (`dot` doesn't.)

The 2D implementation has 4 functions because there are 2 azimuthal coordinate systems and 2 input vectors. The 3D implementation has 6 coordinate systems (raised to the power of) 2 input vectors = 36 functions.

https://github.com/scikit-hep/vector/blob/b9c60a8320b39812354c48182281480087565f5e/src/vector/_compute/planar/dot.py#L44-L49

Four of these are specialized and the others are defined in terms of them.

The 4D implementation has 12 coordinate systems (raised to the power of) 2 input vectors = 144 functions. We don't actually write out those 144 functions; they're generated at startup:

https://github.com/scikit-hep/vector/blob/b9c60a8320b39812354c48182281480087565f5e/src/vector/_compute/lorentz/dot.py#L38-L139

This loop fills the `dispatch_map` with functions, converting all temporal coordinates to Cartesian, but using whatever spatial formulas may be available.

---------------------------

That's not to say that you need to write all of that—I just wanted to make you aware of it so that I can ask, **"In what coordinate systems would it be pragmatic/worthwhile to derive your numerically stable formula?"**
